[master < T938] Bolt v5 support

andrejtonev opened the pull request:
[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Provide the full content or a guide for the final git message


To keep docs changelog up to date, one more thing to do:
- [ ] Write a release note here
- [ ] Tag someone from docs team in the comments

resolves #920 
resolves #921

andrejtonev commented:
Missing notifications_minimum_severity and notifications_disabled_categories.
However, this was not added on purpose, since we are currently ignoring the extra field (as explained in the #939  issue).

gitbuda commented on src/communication/bolt/v1/encoder/base_encoder.hpp:
```diff
@@ -264,6 +295,7 @@ class BaseEncoder {
 
  protected:
   Buffer &buffer_;
+  int major_v_;  //!< Major version of the underlying Bolt protocol (TODO: Think about reimplementing the versioning)
```
Maybe specify what in your view should be changed and how if you have some idea (briefly)

andrejtonev replied:
Added a comment. Basically if we ever refactor the higher-level and end up with classes per bolt version, we could template this and have each bolt version use the correct version out of the box.

antoniofilipovic reviewed (commented):
Looks good, I left few comments. I want to test this on Monday and then that is it

andrejtonev commented:
The current version will fail during the driver tests due to the Java version on the CI server.
Neo4j Java drivers v1 and v4 need Java 11 which is currently installed and the tests should pass.
The v5 drivers need **Java 17**. In addition to Java, the tests use **Maven** package manager (as suggested by the neo4j team) which is missing as well.

Another test that does not fail, but is installing important missing packages on the fly is the GO v5 test. It needs a newer version of golang (>1.18.0), which it downloads and deploys locally. It's not ideal, but is working so far.

@gitbuda

andrejtonev commented on src/communication/bolt/v1/session.hpp:
```diff
@@ -120,6 +120,9 @@ class Session {
         return;
       }
       handshake_done_ = true;
+      // Update the decoder's Bolt version (v5 has changed the undelying structure)
+      decoder_.UpdateVersion(version_.major);
+      encoder_.UpdateVersion(version_.major);
```
Yes, we do not know which version of the protocol we will end up using until the server and client negotiate it. 
We could also destroy the old de/encoder and create a new one that takes in the version, but I wasn't sure if this would have unintended consequences (on the buffer or some other hidden state), so I just added the function call.
I'll add a comment to the code with some ideas about how to improve this part later on.

andrejtonev commented on src/communication/bolt/v1/states/init.hpp:
```diff
@@ -106,6 +114,30 @@ std::optional<Value> GetMetadataV4(TSession &session, const Marker marker) {
+
+template <typename TSession>
+std::optional<Value> GetInitDataV5(TSession &session, const Marker marker) {
+  if (marker != Marker::TinyStruct1) [[unlikely]] {
```
This is defined within the Bolt specifications. The Marker tells you what is incoming. You will se a similar check all over the code. Basically if we receive anything else it is a malformed message.

andrejtonev commented on src/communication/bolt/v1/states/handlers.hpp:
```diff
@@ -465,4 +465,10 @@ State HandleRoute(TSession &session, const Marker marker) {
 }
+
+template <typename TSession>
+State HandleLogOff(TSession &session) {
```
Yeah, or we could just remove it completely. I wasn't sure what will end up in the handler, so I left it just in case. But it seems like it's not needed.

andrejtonev commented on src/communication/bolt/v1/states/executing.hpp:
```diff
@@ -91,6 +91,40 @@ State RunHandlerV4(Signature signature, TSession &session, State state, Marker m
+template <typename TSession>
+State RunHandlerV5(Signature signature, TSession &session, State state, Marker marker) {
+  switch (signature) {
+    case Signature::Run:
+      // Using V4 on purpose
+      return HandleRunV4<TSession>(session, state, marker);
+    case Signature::Pull:
+      // Using V4 on purpose
+      return HandlePullV4<TSession>(session, state, marker);
+    case Signature::Discard:
+      // Using V4 on purpose
+      return HandleDiscardV4<TSession>(session, state, marker);
+    case Signature::Reset:
```
We could :(

andrejtonev commented:
Changelog:

1. Added support for Bolt v5.2
2. Added support for element_id (string representation of ID; currently just mirrors the integer ID)
3. Works with default authorization values (newer Neo4j drivers do not send values that are set to their default values; for example when connecting, the user can skip defining the username and password; this was causing problems with out implementation, but now we prefill the fields to their default values - empty strings in the authorization's case)

@vpavicic

vpavicic commented:
We decided we would not get into details, just say that we support Bolt v5

But this is great info, so up until 5.2, not further?

andrejtonev commented:
Yes. Actually we added support specifically for 5.2. 5.0 and 5.1 are not supported, those will default back to v4
